interop/xds/client/client.go

```diff
@@ -0,0 +1,174 @@
+// LB reporting mechanism that is designed to be queried by an external test
+// driver when verifying that the client is distributing RPCs as expected.
+func (s *statsService) GetClientStats(ctx context.Context, in *testpb.LoadBalancerStatsRequest) (*testpb.LoadBalancerStatsResponse, error) {
+	mu.Lock()
+	watcherKey := statsWatcherKey{currentRequestID, currentRequestID + in.GetNumRpcs()}
+	watcher, ok := watchers[watcherKey]
+	if !ok {
+		watcher = &statsWatcher{
+			rpcsByPeer:    make(map[string]int32),
+			numFailures:   0,
+			remainingRpcs: in.GetNumRpcs(),
+			c:             make(chan testpb.SimpleResponse),
+		}
+		watchers[watcherKey] = watcher
+	}
+	mu.Unlock()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Duration(in.GetTimeoutSec())*time.Second)
+	defer cancel()
+
+	responseCh := make(chan testpb.LoadBalancerStatsResponse)
+	// Wait until the requested RPCs have all been recorded or timeout occurs.
+	go func() {
+		for {
+			select {
+			case r := <-watcher.c:
+				if r.GetServerId() != "" {
+					watcher.rpcsByPeer[r.GetServerId()]++
+				} else {
+					watcher.numFailures++
+				}
+				watcher.remainingRpcs--
+				if watcher.remainingRpcs == 0 {
+					responseCh <- testpb.LoadBalancerStatsResponse{NumFailures: watcher.numFailures + watcher.remainingRpcs, RpcsByPeer: watcher.rpcsByPeer}
+					return
+				}
+			case <-ctx.Done():
+				grpclog.Info("Timed out, returning partial stats")
+				responseCh <- testpb.LoadBalancerStatsResponse{NumFailures: watcher.numFailures + watcher.remainingRpcs, RpcsByPeer: watcher.rpcsByPeer}
+				return
+			}
+		}
+	}()
+
+	response := <-responseCh
+	mu.Lock()
+	delete(watchers, watcherKey)
+	mu.Unlock()
+	return &response, nil
+}
+
+func main() {
+	flag.Parse()
+
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", *statsPort))
+	if err != nil {
+		grpclog.Fatalf("failed to listen: %v", err)
+	}
+	s := grpc.NewServer()
+	defer s.Stop()
+	testpb.RegisterLoadBalancerStatsServiceServer(s, &statsService{})
+	go s.Serve(lis)
+
+	clients := make([]testpb.TestServiceClient, *numChannels)
+	for i := 0; i < *numChannels; i++ {
+		conn, err := grpc.DialContext(context.Background(), *server, grpc.WithInsecure(), grpc.WithBlock())
+		if err != nil {
+			grpclog.Fatalf("Fail to dial: %v", err)
+		}
+		defer conn.Close()
+		clients[i] = testpb.NewTestServiceClient(conn)
+	}
+	ticker := time.NewTicker(time.Second / time.Duration(*qps**numChannels))
+	defer ticker.Stop()
+	sendRpcs(clients, ticker)
+}
+
+func sendRpcs(clients []testpb.TestServiceClient, ticker *time.Ticker) {
+	var i int
+	for range ticker.C {
+		go func(i int) {
+			c := clients[i]
+			ctx, cancel := context.WithTimeout(context.Background(), *rpcTimeout)
+			p := new(peer.Peer)
+			mu.Lock()
+			savedRequestID := currentRequestID
+			currentRequestID++
+			savedWatchers := make(map[statsWatcherKey]*statsWatcher)
+			for key, value := range watchers {
+				savedWatchers[key] = value
```

Comment:
Since this is looping over all the watchers, and `savedRequestID` is known. Why not do the comparison and keep only the interesting channels in a list?

Reply:
Done